fedimint-client/src/db/event_log.rs

```diff
@@ -0,0 +1,499 @@
+    /// useful to get the current count of events.
+    async fn get_next_event_log_id(&mut self) -> EventLogId;
+
+    /// Read a part of the event log.
+    async fn get_event_log(
+        &mut self,
+        pos: Option<EventLogId>,
+        limit: u64,
+    ) -> BTreeMap<
+        EventLogId,
+        (
+            EventKind,
+            Option<(ModuleKind, ModuleInstanceId)>,
+            u64,
+            serde_json::Value,
+        ),
+    >;
+}
+
+#[apply(async_trait_maybe_send!)]
+impl<'tx, Cap> DBTransactionEventLogExt for DatabaseTransaction<'tx, Cap>
+where
+    Cap: Send,
+{
+    async fn log_event_raw(
+        &mut self,
+        log_ordering_wakeup_tx: watch::Sender<()>,
+        kind: EventKind,
+        module_kind: Option<ModuleKind>,
+        module_id: Option<ModuleInstanceId>,
+        payload: Vec<u8>,
+    ) {
+        assert_eq!(
+            module_kind.is_some(),
+            module_id.is_some(),
+            "Events of modules must have module_id set"
+        );
+
+        let unordered_id = UnordedEventLogId::new();
+        trace!(target: LOG_CLIENT_EVENT_LOG, ?unordered_id, "New unordered event log event");
+
+        if self
+            .insert_entry(
+                &unordered_id,
+                &EventLogEntry {
+                    kind,
+                    module: module_kind.map(|kind| (kind, module_id.unwrap())),
+                    ts_msecs: unordered_id.ts_msecs,
+                    payload,
+                },
+            )
+            .await
+            .is_some()
+        {
+            panic!("Trying to overwrite event in the client event log");
+        }
+        self.on_commit(move || {
+            let _ = log_ordering_wakeup_tx.send(());
+        });
+    }
+
+    async fn log_event<E>(
+        &mut self,
+        log_ordering_wakeup_tx: watch::Sender<()>,
+        module_id: Option<ModuleInstanceId>,
+        event: E,
+    ) where
+        E: Event + Send,
+    {
+        self.log_event_raw(
+            log_ordering_wakeup_tx,
+            E::KIND,
+            E::MODULE,
+            module_id,
+            serde_json::to_vec(&event).expect("Serialization can't fail"),
+        )
+        .await;
+    }
+
+    async fn get_next_event_log_id(&mut self) -> EventLogId {
+        self.find_by_prefix_sorted_descending(&EventLogIdPrefixAll)
+            .await
+            .next()
+            .await
+            .map(|(k, _v)| k.next())
+            .unwrap_or_default()
+    }
+
+    async fn get_event_log(
+        &mut self,
+        pos: Option<EventLogId>,
+        limit: u64,
+    ) -> BTreeMap<
+        EventLogId,
+        (
+            EventKind,
+            Option<(ModuleKind, ModuleInstanceId)>,
+            u64,
+            serde_json::Value,
+        ),
+    > {
+        let pos = pos.unwrap_or_default();
+        // TODO: This is innefficent for large logs, and there's no good operations to
+        // help? Do we just query one by one instead?
```

Comment:
This comment is stale already, as I already switched to using range.

Normally it would probably not make much difference, but if for any reason some downstream consumer needs to go through history of event log (tens and hundreds of thousands of items), it might have add up.